### ISSUE TITLE
[material-color-utilities] Dart - Consider respecting SemVer AND pub Dart version conventions

### ISSUE THREAD

**Issue by rydmike** [material-foundation/material-color-utilities#95]
## MaterialColorUtilities Breaks Semantic and Dart Version Conventions

At **two occasions** the package [pub.dev Dart version](https://pub.dev/packages/material_color_utilities) of material_color_utilities has broken **Semantic** and **Dart** pub version conventions:

* **Version 0.1.5** was breaking concerning certain past color results. Using correct version in **Dart pub** conventions, it should have gone from `0.1.4` to `0.2.0`, and not to `0.1.5`.

* **Version 0.3.0** appears to be **none breaking** with only fixes and new features. In Dart pub.dev conventions when using a **0-version package**, the preferred version number from previous `0.2.0`, for a none breaking change, would have been `0.2.1` and not `0.3.0`. If the package has some breaking APIs that are not mentioned in the changelog, and that is the rationale for the breaking version bump, not mentioning the breaking change is not a good practice either. Based on usage findings, it does no appear to have breaking API changes, so first conclusion appears more appropriate. The preferred and less disruptive version number would have been `0.2.1`.

**References** [Semantic versioning](https://semver.org/) and [Dart package versioning](https://dart.dev/tools/pub/versioning):

![Screenshot 2023-04-16 at 19 34 54](https://user-images.githubusercontent.com/39990307/232327053-bad67e59-9ac1-4355-94f9-e0e336119268.png)

## Using O-Version

Since `material_color_utilities` still uses **0-version**, in the Dart **pub** world it means that a change in the **minor** version refers to a **breaking change** and a patch version is a new feature. **Pub.dev** actually resolves **0-version packages** using this **0-version** rule, when it comes to version constraints.

If a package does not respect and abide by these **0-version** Dart pub.dev** rules, it causes complications when other packages try to depend on `material_color_utilities` using correct expectations for a **0-version** package.

### Issue Examples

* `dynamic_color` needed to get a `material_color_utilities` version span bump from `^0.2.0` to `">=0.2.0 <= 0.3.0"`, since its previous constraint did not include `0.3.0`.

![Screenshot 2023-04-16 at 19 50 36](https://user-images.githubusercontent.com/39990307/232327826-55a44608-c227-49f6-9392-c87c0f0184d3.png)

* `flex_seed_scheme` needed to get a `material_color_utilities` version span bump from `^0.2.0` to `">=0.2.0 < 0.4.0"`, since its previous constraint did not include `0.3.0`.

In **both cases** the packages had to update their version constraint for `material_color_utilities` so that the packages can be used on **Flutter stable 3.7.x**, that now uses and is locked to exactly `0.2.0` version **and also on** current **Flutter beta/master 3.10.0-x** versions, that now use and are locked to `0.3.0` version.

#### Risk and complications with this type of version span

It should go without saying that using the above version constraints, that both above examples now **have to use**, goes against good version resolving practices. They now span a bump from **0.2.z** to **0.3.0/0.3.z**, which should refer to an unsafe API version span, but it is not in this case. It is of course also impossible to predict when `material_color_utilities` will break conventions like this, so one cannot include a version span like this in advance.

## Proposal

**Please** consider using package versioning that abide with both **SemVer** and **Dart** conventions.

**Maybe also consider** moving the package away from **0-version** and start calling it `1.y.z`. The conventions are a lot clearer when not using **0-version**. Everybody is more familiar with it, than the above "quirk" on **pub.dev** concerning **0-version** package constraints resolution. 

Since a **0.y.z version** is also in **SemVer** considered to **ONLY** be used: 

> *"For initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable."* 

It is not without concern that **Flutter SDK** depends on a color utility that based on **SemVer** may have totally breaking changes between ANY release. Moving to version `1.y.z` on `material_color_utilities` would help with perceived confidence in the package maturity and its maintenance.

That **Flutter SDK** depends on an external color utility is generally also not without challenges, e.g. when using the package in packages and making them work with different Flutter channels, as seen in above examples.

Moving to version `1.y.z` on `material_color_utilities` the challenge can be partially remedied by never releasing a version that does not offer a minimum shared API that always spans and works from current Flutter **stable**, to **beta** and **master** channels. 

Instead of immediately breaking APIs in `material_color_utilities`, deprecate API that needs to be broken, but initially keep them operational. Add new API, with need new features that could not be added in a none breaking manner.

Bump version from `1.y.z` to `1.y+1.0`. When `1.y+1.0` has landed in **Flutter stable**, as a used **Flutter** dependency, Flutter users and direct package users, get deprecation warning(s), but old API still works, they are urged to, and can move to newer API. When `1.y+1.0` has landed in **Flutter stable**, a new breaking version of `material_color_utilities` can be released, in this example `2.0.0`. All it does is remove the in `1.y+1.0` deprecated, but still operational API. 

Flutter master can then start to use/depend on this version. Sure, this can still have some issues when working with multiple Flutter channels, and packages that also use `material_color_utilities`, but it would be much better and cleaner than the currently messy and challenging situation. 

---

cc/fyi: @guidezpl @esouthren @rodydavis 

---

**EDIT:**  
Edited the statement concerning version 0.3.0 breaking **SemVer**, to recognize @rrousselGit's valid point below, that it is actually not against **SemVer** to make a breaking release, even if there are no breaking changes. To state that it is preferred to not do so.

**Comment by rrousselGit** [comment]
> * **Version 0.3.0** appears to be **none breaking** with only fixes and new features. In Dart pub.dev conventions when using a **0-version package** the correct version number going from previous `0.2.0` would have been `0.2.1` and not `0.3.0`. Either that, or if the package has some breaking APIs it is not mentioned in the changelog, so that would also have been poor practice. Based on my findings, it does no appear to have any breaking API changes so first conclusion is correct, the version number should have been `0.2.1`.

I do not believe that this is correct.   
Packages are free to do """major""" bumps even if there is no real breaking change.   
Semver only seems to be about what the new version should _at minimum_ be. Any version higher than this should be legal.

**Comment by rydmike** [comment]
Ok maybe so. However, if you do that, it causes issues like these without any actual rationale for it, where a none breaking release would have been much less disruptive to users. Why would you prefer to do that? At least I would not consider that good practice, and definitely not nice. Sure there may be cases where it might be motivated, but rarely.

**Comment by guidezpl** [comment]
Thanks for the feedback! I think we erred on the side of safety with `0.3.0` since it contained so many new features, and you're right to point out that `0.1.5` was underbumped given its breaking change. We hope to bump to `1.0.0` soon, where the conventions are clear relative to sub-`1.0.0` versions.

cc @pennzht who owns this library

**Comment by rrousselGit** [comment]
> Why would you prefer to do that?

Riverpod commonly do this.

I try to keep all the Riverpod packages with the same version number more or less.


Meaning that sometimes one of Riverpod packages skip a version. 
Like it may go from 1.0.0 to 1.2.0 directly. 

I wouldn't want flutter_riverpod to be at v3.42.0, but riverpod to be at 2.1.0 if that makes sense.

**Comment by rydmike** [comment]
Thanks @guidezpl, I certainly appreciate the version caution too.

There are indeed a lot of new interesting features in it. Have not had time to try them yet. I plan to, and also to see if I can incorporate some of them in the interactive Themes Playground. 

Since Flutter repo depends on this package too, any breaking change in it makes it hard for packages that depend on this package, to function on different Flutter channels when a Flutter channel switches which version it uses.

Well you see that with your **dynamic_color** package too. I noticed the breaking version constraint span on it too. It had one earlier as well, before it moved to `^0.2.0` and now again to one that spans breaking constraint `">=0.2.0 <= 0.3.0"`. Which in this case works fine, since nothing appears broken in `0.3.0`. Well, as far I have noticed so far anyway and definitely not in my current usage of it. I went with `">=0.2.0 < 0.4.0"`, in case there are any feature releases to `0.3.0`. It still feels a bit unsound to use these kind of constraints, but at least then my package users can use it on all current Flutter channels.

Looking forward to the 1.0.0 release 👍🏻 It is indeed much clearer than 0-ver, plus it imo sends a better message. Which I think is important since Flutter stable depends on the package.

I think my unintentionally maybe a bit harsh sounding feedback, has been noted, so I'm closing this issue 😄💙
